pipelinewise/fastsync/commons/tap_mysql.py

```diff
@@ -207,15 +212,37 @@ def copy_table(self, table_name, path):
                     if not rows:
                         break
 
-
                     # Log export status
                     exported_rows += len(rows)
-                    if (len(rows) == export_batch_rows):
-                        #Then we believe this to be just an interim batch and not the final one so report on progress
-                        utils.log("Exporting batch from {} to {} rows from {}...".format((exported_rows-export_batch_rows),exported_rows, table_name))
+                    if len(rows) == export_batch_rows:
+                        # Then we believe this to be just an interim batch and not the final one so report on progress
+                        utils.log(
+                            "Exporting batch from {} to {} rows from {}...".format((exported_rows - export_batch_rows),
+                                                                                   exported_rows, table_name))
 
                     # Write rows to file
                     for row in rows:
+                        row = list(row)
+                        for ind, val in enumerate(row):
+                            if isinstance(val, bytes):
+                                # Removes all trailing '\x00' from the bytes Trailing zeroes in bytes is caused by
+                                # how BINARY type columns pads the inserted values, this doesn't happen when using
+                                # VARBINARY.
+                                #
+                                # ref: https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
+                                #
+                                # Leaving the trailing zeroes makes FT & Incremental behave differently than Binlog
+                                # and we end up with inconsistent state records.
+                                #
+                                # Example: for a binary value b'pk0'
+                                #   - Binlog would read it as b'pk0' then in HEX it would be b'706b30'
+                                #   - FT & Inc would read it as b'pk0\x00\x00\x00\x00....\x00' then in HEX
+                                #       it would be b'706b300000...0'
+                                stripped_utf8_elem = val.decode().rstrip('\x00').encode()
```

Comment:
The same code as in https://github.com/transferwise/pipelinewise-tap-mysql/pull/9
So same concerns apply here :)

Reply:
is this the same functionality to the previously used mysql `HEX(...)` function that we do at export time at https://github.com/transferwise/pipelinewise/pull/293/files/a76d4e4609350f89b44a572c202c9d37b97dfe6b#diff-8086c1c0711f0b49a47a6c60db4e2260R150?

If yes, maybe we can leave it to MySQL because these operations usually faster doing in mysql than doing it by python.

Reply:
> The same code as in [transferwise/pipelinewise-tap-mysql#9](https://github.com/transferwise/pipelinewise-tap-mysql/pull/9)
> So same concerns apply here :)

Check my comments in https://github.com/transferwise/pipelinewise-tap-mysql/pull/9



> is this the same functionality to the previously used mysql `HEX(...)` function that we do at export time [here](https://github.com/transferwise/pipelinewise/pull/293/files/a76d4e4609350f89b44a572c202c9d37b97dfe6b#diff-8086c1c0711f0b49a47a6c60db4e2260R149) ?
> 
> If yes, maybe we can leave it to MySQL because these operations usually faster doing in mysql than doing it by python.

I have thought of that, Hex is not enough to trim the extra 0s, I have spent 2-3 days looking for ways to trim using sql but nothing worked. So doing it using code was my last option. If you know how to do it please do share.

Reply:
We can't trim after using hex because it would also trim zeros that are part of the data.